Background Request Handler

sanderpick opened the pull request:
This updates mobile to 0.4.1 and includes an implementation of the `CafeOutboxHandler`.

- [x] Properly block in `flush`
- [x] Include request events in the messenger
- [x] Write a test

sanderpick commented:
Made some good progress here today. I started w/ a sync test, but ended up moving to a full Textile test because the mocking was getting nightmarish. Will wrap that up minimally tomorrow. The aim here is not too get full Textile coverage in this PR.

asutula commented on textile/src/main/java/io/textile/textile/MessageHandler.java:
```diff
@@ -3,121 +3,161 @@
-                        break;
-                    default:
-                        break;
+        switch (MobileEventType.forNumber(event.getType())) {
```
Nice, didn't know this event type enum exists

asutula commented:
Looking good so far. We should sync up and keep our general architecture aligned between platforms.

asutula commented on textile/build.gradle:
```diff
@@ -23,28 +23,46 @@ ext {
+    Properties properties = new Properties()
+    properties.load(project.rootProject.file('local.properties').newDataInputStream())
+    return properties.getProperty("cafe.token")
+}
```
Better to use env vars, easier with CI. See an example in this file like:
```
String.valueOf(System.getenv("BINTRAY_USERNAME"))
```

sanderpick replied:
I tried this for like an hour yesterday... couldn't get it to pick up the system env value.

sanderpick replied:
Oh, I see what you're saying... in the gradle file. Will try that, thanks 👊

sanderpick commented:
This is all updated with mobile v0.5.0. Couple things left to do:
- [ ] Ensure flush is called on a single thread
- [ ] Fix broadcast receiver registration

sanderpick commented:
Updates:
- Flush now uses a single thread executor queue. Inside the queue, batch of request ids are handled in parallel with `CompletableFutures` - which leverage a different thread pool. The effect of handling the disc writes and uploading queueing for a batch in parallel is really noticeable. I'm pretty happy with how that came together. However, in order to user the nice concurrency libs, I had to:
  - Set min sdk to 24
  - Switched to Java8
- AS kept insisting an update to gradle 5, so far so good
- Now for the broadcaster receiver stuff... in theory very simple, but after wasting a _bunch_ of time tinkering, I have to ask you to take a look at this @asutula and see if you can understand why the broadcaster receiver is not receiving events. It would be easiest to hop on a call and walk through it. I've tried everything I can think of... spinning wheels. Need some fresh eyes because it's likely something simple.

The demo app is currently wired up to register with dev, create a thread, and add some data, which will kick off the requests. You can see they all complete, but we only get the initial status event from marking each one as pending.

By way of trying to get this to work, I've also:
- Updated to the latest uploader lib, which forced an update to androidx
- Got tired of tinkering with versioning between `app` and `textile`, so created a `manifest.gradle` which holds all env vars

Refs:
- Overview of monitoring upload status with the uploader lib: https://github.com/gotev/android-upload-service/wiki/Monitoring-upload-status (note that a broadcast receiver is needed to continue to receive events when the activity is destroyed, which is critical for our implementation, as we have to mark them as complete so the blocks can be released.
- Tried this: https://github.com/gotev/android-upload-service/wiki/Monitoring-upload-status#broadcast-receiver-inside-service-or-activity- (current check-in status)
- And this: https://github.com/gotev/android-upload-service/wiki/Monitoring-upload-status#global-broadcast-receiver- (including receiver in lib and/or demo app manifest)
- Uploader lib example: https://github.com/gotev/android-upload-service/tree/master/examples/app/demoapp
- PR on the RN uploader that also uses the gotev lib and supposedly is working (received broadcasted events) https://github.com/Vydia/react-native-background-upload/pull/133/files

sanderpick commented:
Easiest way to set the cafe id and token locally is to add this to your `local.properties` file:
```
test.cafe.id=12D3KooWGN8VAsPHsHeJtoTbbzsGjs2LTmQZ6wFKvuPich1TYmYY
test.cafe.token=uggU4NcVGFSPchULpa2zG2NRjw2bFzaiJo3BYAgaFyzCUPRLuAgToE3HXPyo
```
If those are not present, the build config will look for `TEST_CAFE_ID` and `TEST_CAFE_TOKEN` env vars.

sanderpick commented:
Okay, I might be on to something here... more after dinner.

sanderpick commented:
At last, sync events are coming through the broadcast receiver.

![DSBXVVk](https://user-images.githubusercontent.com/361000/59994239-fd4b0380-9607-11e9-82fe-57c5f2e8088a.gif)

I ended up narrowing it all down to one line in the uploader lib. I think the issue is that it's not meant to be easily embedded in another library. Forked and removed: https://github.com/textileio/android-upload-service/commit/bca1d1af94b351be2d79bf9caa9b95c9a83c5d3c#diff-1fbd837a670d236e4986465b5871851aL33 (using https://jitpack.io/ to pull it in as a maven dependency)

```
I/System.out: ------------> sync update QmZtzxU49uaZykgvqLxwqm1mxSz7fEP5aHJWvtf7GyQtST: 0%
I/System.out: ------------> sync update QmZtzxU49uaZykgvqLxwqm1mxSz7fEP5aHJWvtf7GyQtST: 13%
I/System.out: ------------> sync update QmZtzxU49uaZykgvqLxwqm1mxSz7fEP5aHJWvtf7GyQtST: 15%
I/System.out: ------------> sync update QmZtzxU49uaZykgvqLxwqm1mxSz7fEP5aHJWvtf7GyQtST: 25%
.
.
.
I/System.out: ------------> sync complete: QmZtzxU49uaZykgvqLxwqm1mxSz7fEP5aHJWvtf7GyQtST
```

asutula commented:
Cool thanks for the update. I'm glad things are coming together, but also a little curious about why some of those things needed to be done and wondering what the consequences will be. I'm no Android expert, so we need to do some research and testing to see what the move to min sdk 24, Java 8 and Android X do to a user of the SDK. I'm pretty sure its going to be a problem, especially the Android X dependency. We previously had a dependency of the Photos app sneak up to Android X and it broke our compilation. Seems hard or impossible to mix pre-Android X and Android X libs in a project... Some "mainfest merge" step fails. I've also been keeping up with the React Native 0.6 RC release notes where they are moving to Android X and describing it as a breaking change.

Let's hop on a call tonight (the morning for you) and go through all this. Good work!

sanderpick commented:
Cool, sounds good. Remember we’re in a hurry - let’s just try this in the sdk ASAP.. You can merge manifests with androidx. I was doing that for awhile. 

I don’t think the fix in the end is related to having the latest uploader lib- so if there are problems we can revert for a bit. There’s a bunch of improvement in here though that would be nice to retain.

sanderpick commented:
I got painfully close to keeping androidx last night using https://www.npmjs.com/package/jetifier and react-native 0.60-RC, (see convo here: https://github.com/react-native-community/discussions-and-proposals/issues/129#issuecomment-504169271), but the app would just crash immediately and it seemed just a little too precarious to keep pushing down that path.

I ended up un-migrating the upload lib to androidx this morning: https://github.com/textileio/android-upload-service/commit/7ccec4c91b1e9fceb57dd9bdd101ce21f992946a

Running in the photos app. Have a look here @asutula, let's get this in.

sanderpick commented:
Sweet, thanks for the review! I made one last change here: Handle the upload broadcast events with single executor. I'm still not sure if the app can get into a state where the single `Textile` instance does not exist when the events come in. Based on what I've read so far, since we're registering the broadcast receiver in the manifest...

> The difference of registering BroadcastReceivers with Context.registerReceiver() is that you must eventually also unregister it, while receivers defined in you manifest will wake up your app no matter what. When to unregister depends on where you first register your receiver. (https://medium.com/@iiro.krankka/its-time-to-kiss-goodbye-to-your-implicit-broadcastreceivers-eefafd9f4f8a)

The app should be awake, but I'm not sure if that means the `Lifecycle` service would have necessarily created `Textile` yet. Does the app get "launched into the background", like on iOS? In any case, let's play around with this and see if some extra handling needs to be added that ensures `Textile.instance()` is not `null`.

sanderpick commented:
> Awesome. Cool have Java 8 at our disposal now! Will have to expand the use of those futures.

Totally! Already used in the photos app's native `TextileImageTask` 💎
